libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -130,24 +136,46 @@ void AC_WPNav::wp_and_spline_init()
     // initialise position controller
     _pos_control.set_desired_accel_xy(0.0f,0.0f);
     _pos_control.init_xy_controller();
-    _pos_control.clear_desired_velocity_ff_z();
 
     // initialise feed forward velocity to zero
     _pos_control.set_desired_velocity_xy(0.0f, 0.0f);
 
     // initialize the desired wp speed if not already done
-    _wp_desired_speed_xy_cms = _wp_speed_cms;
+    _wp_desired_speed_xy_cms = is_positive(speed_cms) ? speed_cms : _wp_speed_cms;
 
     // initialise position controller speed and acceleration
-    _pos_control.set_max_speed_xy(_wp_speed_cms);
+    _pos_control.set_max_speed_xy(_wp_desired_speed_xy_cms);
     _pos_control.set_max_accel_xy(_wp_accel_cmss);
     _pos_control.set_max_speed_z(-_wp_speed_down_cms, _wp_speed_up_cms);
     _pos_control.set_max_accel_z(_wp_accel_z_cmss);
     _pos_control.calc_leash_length_xy();
     _pos_control.calc_leash_length_z();
 
-    // initialise yaw heading to current heading target
+    float jerk = MIN(_attitude_control.get_ang_vel_roll_max() * GRAVITY_MSS, _attitude_control.get_ang_vel_pitch_max() * GRAVITY_MSS);
+    if (is_zero(jerk)) {
+        jerk = _wp_jerk;
+    } else {
+        jerk = MIN(jerk, _wp_jerk);
+    }
```

Comment:
This is a bit funny because we take a jerk parameter from the user and then we make sure we can achieve that jerk based on the attitude controller limits. Our standard for these parameters is they are disabled if set to zero. We do this because if we actually set a limit of zero for jerk, acceleration, or velocity then the aircraft would never move so it is safe to say if they are zero then we are not applying that limit.

So in this case we are really saying we want the lowest non-zero value based on the limits set by the user. Then if they are all set to zero then we will set it to 0.1f.

Maybe 0.1f should be some sort of #define WPNAV_JERK 0.1f